### PR TITLE
[Performance] Support both xgrammar and outlines for constrained decoding

### PR DESCRIPTION
**Pull request by DarkSharpness**
## Motivation

This pull request is a continuation of [PR #1680](https://github.com/sgl-project/sglang/pull/1680). 

This pull request allows users to choose between `xgrammar` and `outlines`.

## Modifications

Key modifications include:

- Added support for both `xgrammar` and `outlines` as grammar backends.
- Users can select their preferred backend by passing the `grammar-backend` option in the command line.
- The default backend is set to `outlines`.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/managers/schedule_batch.py [hunk 1]
```diff
@@ -247,8 +247,11 @@ def __init__(
         self.embedding = None
 
         # Constrained decoding
-        self.regex_fsm: RegexGuide = None
+        self.regex_fsm: Optional[RegexGuide] = None
         self.regex_fsm_state: int = 0
+        self.regex_bnf: Optional[GrammarMatcher] = None
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py]: Build a better abstraction for this.
Merge `regex_fsm`, `regex_fsm_state`, `regex_bnf` into one variable and the scheduler should not know it is outlines or xgrammar.

File: python/sglang/srt/managers/schedule_batch.py [hunk 2]
```diff
@@ -410,6 +413,72 @@ def jump_forward_and_retokenize(self, jump_forward_str, next_state):
 
         return True
 
+    def jump_forward_and_retokenize_bnf(self, jump_forward_str):
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py]: Build a better abstraction
You can build a new class called `GrammarBackend`. The scheduler just talks to `GrammarBackend` without knowing it is outlines/xgrammar/fsm/bnf..

File: python/sglang/srt/managers/schedule_batch.py [hunk 3]
```diff
@@ -247,8 +247,11 @@ def __init__(
         self.embedding = None
 
         # Constrained decoding
-        self.regex_fsm: RegexGuide = None
+        self.regex_fsm: Optional[RegexGuide] = None
         self.regex_fsm_state: int = 0
+        self.regex_bnf: Optional[GrammarMatcher] = None
+
+        self.allow_jump_forward = False
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py]: turn on this?

File: python/sglang/srt/managers/scheduler.py [hunk 4]
```diff
@@ -224,16 +225,31 @@ def __init__(
...
+            elif server_args.grammar_backend == "outlines":
+                self.regex_fsm_cache = FSMCache(
+                    server_args.tokenizer_path,
+                    {
+                        "tokenizer_mode": server_args.tokenizer_mode,
+                        "trust_remote_code": server_args.trust_remote_code,
+                    },
+                    skip_tokenizer_init=server_args.skip_tokenizer_init,
+                    constrained_json_whitespace_pattern=server_args.constrained_json_whitespace_pattern,
+                )
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]: this should be the following without a branch
```
self.grammar_cache = GrammarCache( ..., backend=server_args.grammar_backend)
```

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/managers/schedule_batch.py (hunk 3 above)
Comment (DarkSharpness) [reply python/sglang/srt/managers/schedule_batch.py]: Thank you for your feedback!

We've carefully reviewed and implemented the suggested changes. Specifically, we've now built an abstraction of `Grammar` to handle constrained decoding more elegantly.

**Review by merrymercy — APPROVED** [review]

**Pull request merged**